contracts/staking/ValidatorShare.sol

```diff
@@ -0,0 +1,157 @@
+  }
+
+  function buyVoucher(uint256 _amount) public onlyWhenUnlocked {
+    uint256 share = _amount.mul(100).div(exchangeRate());
+    totalStake = totalStake.add(_amount);
+    amountStaked[msg.sender] = amountStaked[msg.sender].add(_amount);
+    require(token.transferFrom(msg.sender, address(this), _amount), "Transfer amount failed");
+    _mint(msg.sender, share);
+    activeAmount = activeAmount.add(_amount);
+    IStakeManager(owner()).updateValidatorState(validatorId, int256(_amount));
+
+    stakingLogger.logShareMinted(validatorId, msg.sender, _amount, share);
+    stakingLogger.logStakeUpdate(validatorId);
+  }
+
+  function sellVoucher() public {
+    uint256 share = balanceOf(msg.sender);
+    require(share > 0, "Zero balance");
+    uint256 _amount = exchangeRate().mul(share).div(100);
+    _burn(msg.sender, share);
+    IStakeManager(owner()).updateValidatorState(validatorId, -int256(_amount));
+
+    IStakeManager stakeManager = IStakeManager(owner());
+    activeAmount = activeAmount.sub(_amount);
+    if (_amount > amountStaked[msg.sender]) {
+      uint256 _rewards = _amount.sub(amountStaked[msg.sender]);
+      //withdrawTransfer
+      IStakeManager(owner()).delegationTransfer(validatorId, _rewards, msg.sender);
+      _amount = _amount.sub(_rewards);
+    }
+
+    amountStaked[msg.sender] = 0; // TODO: add partial sell amountStaked[msg.sender].sub(_amount);
+    delegators[msg.sender] = Delegator({
+        amount: _amount,
+        withdrawEpoch: stakeManager.currentEpoch().add(stakeManager.WITHDRAWAL_DELAY())
+      });
+
+    stakingLogger.logShareBurned(validatorId, msg.sender, _amount, share);
+    stakingLogger.logStakeUpdate(validatorId);
+  }
+
+  function withdrawRewards() public {
+    uint256 liquidRewards = getLiquidRewards(msg.sender);
+    uint256 sharesToBurn = liquidRewards.mul(100).div(exchangeRate());
+    // if (sharesToBurn > 0)
+    _burn(msg.sender, sharesToBurn);
+    rewards = rewards.sub(liquidRewards);
+    IStakeManager(owner()).delegationTransfer(validatorId, liquidRewards, msg.sender);
+    stakingLogger.logDelClaimRewards(validatorId, liquidRewards, sharesToBurn);
+  }
+
+  function reStake() public {
+    /**
+      - only active amount is considers as active stake
+      - move reward amount to active stake pool
+      - no shares are minted
+     */
+    uint256 liquidRewards = getLiquidRewards(msg.sender);
+    amountStaked[msg.sender] = amountStaked[msg.sender].add(liquidRewards);
+    totalStake = totalStake.add(liquidRewards);
+    activeAmount = activeAmount.add(liquidRewards);
+    IStakeManager(owner()).delegationTransfer(validatorId, liquidRewards, address(this));
+    IStakeManager(owner()).updateValidatorState(validatorId, int256(liquidRewards));
+    rewards = rewards.sub(liquidRewards);
+    stakingLogger.logStakeUpdate(validatorId);
+    // TODO: add restake event
+  }
+
+  function getLiquidRewards(address user) internal returns(uint256 liquidRewards) {
```

Comment:
Mark this function as view and please pay attention to compiler warnings - they help the code to be readable and prevent bugs in the future.

Reply:
Thanks but what's the use of internal view functions, other than restricting that it's not updating any state.

Reply:
You answered your own question - it restricts state changes. And is much more readable.

Reply:
fixed